homeassistant/components/growatt/sensor.py

```diff
@@ -0,0 +1,178 @@
+            hass, growatt_client, username, password
+        )
+        sensor_current = GrowattPlantCurrent(
+            hass, growatt_client, username, password
+        )
+        add_entities([sensor_today, sensor_total, sensor_current])
+        return True
+
+    return False
+
+
+def login(client, username, password):
+    """Login to the growatt server."""
+    import growatt
+
+    try:
+        client.login(username, password)
+        return True
+    except growatt.LoginError as error:
+        logging.error(error)
+        return False
+
+
+class GrowattPlant(Entity):
+    """Base Growatt sensor class."""
+
+    def __init__(self, hass, client, username, password):
+        """Initialize the sensor."""
+        self._hass = hass
+        self._client = client
+        self._username = username
+        self._password = password
+        self._state = None
+
+    @property
+    def state(self):
+        """Return the state of the sensor."""
+        return self._state
+
+    @staticmethod
+    def convert_multiplier(metric_name, multiplier_lookup):
+        """Convert a value to a given multiplier."""
+        if metric_name not in multiplier_lookup:
+            message = (
+                "Found an unsupported metric name {}"
+                "cannot convert safely to kWh."
+            ).format(metric_name)
+            logging.error(message)
+            raise ValueError(message)
+        return multiplier_lookup[metric_name]
+
+
+class GrowattPlantTotals(GrowattPlant):
+    """Representation of a Growatt plant sensor."""
+
+    @property
+    def unit_of_measurement(self):
+        """Return the unit this state is expressed in."""
+        return "kWh"
+
+    def _convert_to_kwh(self, value: str, metric_name: str):
+        """Convert a value to a kWh value."""
+        watts = float(value)
+        multiplier_lookup = {"kWh": 1, "MWh": 1000, "GWh": 1000 * 1000}
+        return watts * self.convert_multiplier(metric_name, multiplier_lookup)
+
+    def _get_total_energy(self, key: str):
+        """Get todays energy as float in kWh.
+
+        Refreshes login to update the session.
+        """
+        login(self._client, self._username, self._password)
+        plant_info = self._client.plant_list()
+        return float(
+            self._convert_to_kwh(*plant_info["totalData"][key].split(" "))
+        )
+
+
+class GrowattPlantToday(GrowattPlantTotals):
+    """Representation of a Growatt plant daily sensor."""
+
+    def update(self):
+        """Get the latest data from Growatt server."""
+        self._state = self._get_total_energy("todayEnergySum")
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return "Growatt plant today"
+
+
+class GrowattPlantTotal(GrowattPlantTotals):
+    """Representation of a Growatt plant total sensor."""
+
+    def update(self):
+        """Get the latest data from Growatt server."""
+        self._state = self._get_total_energy("totalEnergySum")
```

Comment:
Only different for today and total sensor is metric name, I suggest you define a instance var `self._metric_name` in base class, then you don't need define two subclass.

I think this also apply to current sensor, they just have different unit of measurement, you should be able just use one class for all of your sensors.

Reply:
I've merged totals and today sensors into one class. The other one is significant different that it justifies a subclass.